{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d80061ff-7866-42d2-8dca-6dfebbbc65c1",
   "metadata": {},
   "source": [
    "Source: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE145926  \n",
    "Title\tSingle-cell landscape of bronchoalveolar immune cells in COVID-19 patients  \n",
    "Organism\tHomo sapiens  \n",
    "Experiment type\tExpression profiling by high throughput sequencing  \n",
    "Other  \n",
    "Summary\tImmune characteristics associated with Coronavirus Disease-2019 (COVID-19) severity are currently unclear. We characterized bronchoalveolar lavage fluid (BALF) immune cells from patients with varying severity of COVID-19 disease and from healthy subjects using single-cell RNA-sequencing. Proinflammatory monocyte-derived macrophages were abundant in the BALF from severe COVID-9 patients. Moderate cases were characterized by the presence of highly clonally expanded tissue-resident CD8+ T cells. This atlas of the bronchoalveolar immune-microenvironment suggests potential mechanisms underlying pathogenesis and recovery in COVID-19.  \n",
    " \t\n",
    "Overall design\tUsing 10x genomics to measure single-cell RNA sequence (scRNA-seq)/TCR-seq to comprehensively characterize the lung immune microenvironment in the bronchoalveolar lavage fluid (BALF) from 6 severe and 3 moderate COVID-19 patients and 3 healthy control.  \n",
    "\n",
    "GSM4339769: mild COVID-19  \n",
    "GSM4339770: mild COVID-19  \n",
    "GSM4339771: severe COVID-19  \n",
    "GSM4339772: mild COVID-19  \n",
    "GSM4339773: severe COVID-19  \n",
    "GSM4339774: severe COVID-19  \n",
    "GSM4475048: control  \n",
    "GSM4475049: control  \n",
    "GSM4475050: control  \n",
    "GSM4475051: control  \n",
    "GSM4475052: control  \n",
    "GSM4475053: control  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46df702e-e287-40b8-8d7b-a3eaeccca3be",
   "metadata": {},
   "source": [
    "## Part I: Exploratory & Seperation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "393bc8e1-5c4c-461e-83f3-7aca349048f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import h5py\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4bf81e6b-3c96-4130-9967-07ecc248449e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import matplotlib\n",
    "import sklearn\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from sklearn import preprocessing\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.decomposition import PCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "82baeea5-49ce-4017-a114-90107e9713af",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['file_description.rtf',\n",
       " 'GSM4475052_C149_filtered_feature_bc_matrix.h5',\n",
       " 'GSM4339774_C146_filtered_feature_bc_matrix.h5',\n",
       " 'GSM4475049_C52_filtered_feature_bc_matrix.h5',\n",
       " 'GSM4339772_C144_filtered_feature_bc_matrix.h5',\n",
       " 'GSM4475053_C152_filtered_feature_bc_matrix.h5',\n",
       " 'GSM4475051_C148_filtered_feature_bc_matrix.h5',\n",
       " 'GSM4339770_C142_filtered_feature_bc_matrix.h5',\n",
       " 'GSM4339769_C141_filtered_feature_bc_matrix.h5',\n",
       " 'GSM4339771_C143_filtered_feature_bc_matrix.h5',\n",
       " 'GSM4475048_C51_filtered_feature_bc_matrix.h5',\n",
       " 'GSM4339773_C145_filtered_feature_bc_matrix.h5',\n",
       " 'GSM4475050_C100_filtered_feature_bc_matrix.h5',\n",
       " '.ipynb_checkpoints',\n",
       " 'file_description.txt',\n",
       " 'ML_Script_1_data_preprocessing.ipynb',\n",
       " 'all.cell.annotation.meta.txt']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.listdir()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3ccdea78-2f23-4c3e-8a2d-02b5a64f6c58",
   "metadata": {},
   "outputs": [],
   "source": [
    "mild_COVID = ['GSM4339769', 'GSM4339770', 'GSM4339772']\n",
    "severe_COVID = ['GSM4339771', 'GSM4339773', 'GSM4339774']\n",
    "control_ = ['GSM4475048', 'GSM4475049', 'GSM4475050', 'GSM4475051', 'GSM4475052', 'GSM4475053']\n",
    "\n",
    "mild = []\n",
    "severe = []\n",
    "control = []\n",
    "\n",
    "for f in os.listdir():\n",
    "    if f.endswith('.h5'):\n",
    "        for names in mild_COVID:\n",
    "            if f.startswith(names):\n",
    "                mild.append(f)\n",
    "\n",
    "        for names in severe_COVID:\n",
    "            if f.startswith(names):\n",
    "                severe.append(f)\n",
    "\n",
    "        for names in control_:\n",
    "            if f.startswith(names):\n",
    "                control.append(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4d1da30b-19e1-4146-b5be-2fd2e53d770f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mild\n",
      "['GSM4339772_C144_filtered_feature_bc_matrix.h5', 'GSM4339770_C142_filtered_feature_bc_matrix.h5', 'GSM4339769_C141_filtered_feature_bc_matrix.h5']\n",
      "severe\n",
      "['GSM4339774_C146_filtered_feature_bc_matrix.h5', 'GSM4339771_C143_filtered_feature_bc_matrix.h5', 'GSM4339773_C145_filtered_feature_bc_matrix.h5']\n",
      "control\n",
      "['GSM4475052_C149_filtered_feature_bc_matrix.h5', 'GSM4475049_C52_filtered_feature_bc_matrix.h5', 'GSM4475053_C152_filtered_feature_bc_matrix.h5', 'GSM4475051_C148_filtered_feature_bc_matrix.h5', 'GSM4475048_C51_filtered_feature_bc_matrix.h5', 'GSM4475050_C100_filtered_feature_bc_matrix.h5']\n"
     ]
    }
   ],
   "source": [
    "print('mild')\n",
    "print(mild)\n",
    "print('severe')\n",
    "print(severe)\n",
    "print('control')\n",
    "print(control)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "890f539f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# with h5py.File(os.path.join(os.getcwd(), 'GSM4339769_C141_filtered_feature_bc_matrix.h5'), 'r') as f:\n",
    "#     # list all slots\n",
    "#     print(list(f.keys()))\n",
    "#     print(f['matrix'])\n",
    "\n",
    "#     # List all groups (keys)\n",
    "#     print(\"Keys in 'matrix' group:\")\n",
    "#     for key in f['matrix'].keys():\n",
    "#         print(key)\n",
    "        \n",
    "#     for key, value in f['matrix'].items():\n",
    "#         print(f\"{key}: {value}\")\n",
    "#         print(f\"{value}.shape\")\n",
    "\n",
    "# # Read the data (testing)\n",
    "# with h5py.File(os.path.join(os.getcwd(), 'GSM4339769_C141_filtered_feature_bc_matrix.h5'), 'r') as f:\n",
    "#     # Read data and labels\n",
    "#     data = f['matrix']['data'][()]  # Read 'data' dataset\n",
    "#     barcodes = f['matrix']['barcodes'][()]  # Read 'barcodes' dataset\n",
    "#     features = f['matrix']['features']\n",
    "#     indices = f['matrix']['indices']\n",
    "#     indptr = f['matrix']['indptr']\n",
    "#     shape = f['matrix']['shape']\n",
    "\n",
    "#     print(data[0:5])\n",
    "#     print(barcodes[0:5])\n",
    "#     print(features)\n",
    "#     print(indices)\n",
    "#     print(indptr)\n",
    "#     print(shape)\n",
    "\n",
    "#     # Get the number of features and barcodes\n",
    "#     num_features = shape[0]\n",
    "#     num_barcodes = shape[1]\n",
    "    \n",
    "    # Create a sparse matrix using the CSR (Compressed Sparse Row) format\n",
    "#     count_matrix = np.zeros((num_features, num_barcodes), dtype=int)\n",
    "\n",
    "#     for i in range(num_barcodes):\n",
    "#         barcode_data = data[indptr[i]:indptr[i+1]]\n",
    "#         barcode_indices = indices[indptr[i]:indptr[i+1]]\n",
    "#         count_matrix[barcode_indices, i] = barcode_data\n",
    "\n",
    "#     # Create a DataFrame from the count matrix\n",
    "#     count_df = pd.DataFrame(data=count_matrix, columns=f['matrix']['barcodes'][:])\n",
    "\n",
    "# # Display the count matrix DataFrame\n",
    "# print(count_df)\n",
    "\n",
    "# Create a DataFrame from the data\n",
    "# df_test = pd.DataFrame(data=)\n",
    "\n",
    "# Display the data in the DataFrame\n",
    "# print(df_test.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fdb06081-f248-472a-8af7-4a02b112bedb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<HDF5 group \"/matrix/features\" (5 members)>\n",
      "Keys in 'features' group:\n",
      "\n",
      "Dataset Name: _all_tag_keys\n",
      "Shape: (1,)\n",
      "Type: |S6\n",
      "\n",
      "Dataset Name: feature_type\n",
      "Shape: (33539,)\n",
      "Type: |S15\n",
      "\n",
      "Dataset Name: genome\n",
      "Shape: (33539,)\n",
      "Type: |S11\n",
      "\n",
      "Dataset Name: id\n",
      "Shape: (33539,)\n",
      "Type: |S15\n",
      "\n",
      "Dataset Name: name\n",
      "Shape: (33539,)\n",
      "Type: |S16\n"
     ]
    }
   ],
   "source": [
    "with h5py.File(os.path.join(os.getcwd(), 'GSM4339769_C141_filtered_feature_bc_matrix.h5'), 'r') as f:\n",
    "    # Print the HDF5 group and its members\n",
    "    print(f['matrix/features'])\n",
    "    print(\"Keys in 'features' group:\")\n",
    "    for key in f['matrix/features'].keys():\n",
    "        print(\"\\nDataset Name:\", key)\n",
    "        dataset = f['matrix/features'][key]\n",
    "        print(\"Shape:\", dataset.shape)\n",
    "        print(\"Type:\", dataset.dtype)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e93e63f1-1347-4095-8df4-c704eeae3ea9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAACCTGAGATGTCGG-1</th>\n",
       "      <th>AAACCTGAGGCTCATT-1</th>\n",
       "      <th>AAACCTGAGTTGTCGT-1</th>\n",
       "      <th>AAACCTGCAATCCGAT-1</th>\n",
       "      <th>AAACCTGCAGGCGATA-1</th>\n",
       "      <th>AAACCTGCATGGTCAT-1</th>\n",
       "      <th>AAACCTGGTACCGTAT-1</th>\n",
       "      <th>AAACCTGGTTAAGGGC-1</th>\n",
       "      <th>AAACCTGGTTTAGCTG-1</th>\n",
       "      <th>AAACCTGTCAATCACG-1</th>\n",
       "      <th>...</th>\n",
       "      <th>TTTGTCAAGTGGTAAT-1</th>\n",
       "      <th>TTTGTCAAGTGTTGAA-1</th>\n",
       "      <th>TTTGTCACAAGAAGAG-1</th>\n",
       "      <th>TTTGTCACAGAAGCAC-1</th>\n",
       "      <th>TTTGTCAGTAGGGACT-1</th>\n",
       "      <th>TTTGTCAGTTCAGGCC-1</th>\n",
       "      <th>TTTGTCATCAACCAAC-1</th>\n",
       "      <th>TTTGTCATCACATGCA-1</th>\n",
       "      <th>TTTGTCATCCAAACAC-1</th>\n",
       "      <th>TTTGTCATCGCGTTTC-1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 6249 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             AAACCTGAGATGTCGG-1  AAACCTGAGGCTCATT-1  AAACCTGAGTTGTCGT-1  \\\n",
       "MIR1302-2HG                   0                   0                   0   \n",
       "FAM138A                       0                   0                   0   \n",
       "OR4F5                         0                   0                   0   \n",
       "AL627309.1                    0                   0                   0   \n",
       "AL627309.3                    0                   0                   0   \n",
       "\n",
       "             AAACCTGCAATCCGAT-1  AAACCTGCAGGCGATA-1  AAACCTGCATGGTCAT-1  \\\n",
       "MIR1302-2HG                   0                   0                   0   \n",
       "FAM138A                       0                   0                   0   \n",
       "OR4F5                         0                   0                   0   \n",
       "AL627309.1                    0                   0                   0   \n",
       "AL627309.3                    0                   0                   0   \n",
       "\n",
       "             AAACCTGGTACCGTAT-1  AAACCTGGTTAAGGGC-1  AAACCTGGTTTAGCTG-1  \\\n",
       "MIR1302-2HG                   0                   0                   0   \n",
       "FAM138A                       0                   0                   0   \n",
       "OR4F5                         0                   0                   0   \n",
       "AL627309.1                    0                   0                   0   \n",
       "AL627309.3                    0                   0                   0   \n",
       "\n",
       "             AAACCTGTCAATCACG-1  ...  TTTGTCAAGTGGTAAT-1  TTTGTCAAGTGTTGAA-1  \\\n",
       "MIR1302-2HG                   0  ...                   0                   0   \n",
       "FAM138A                       0  ...                   0                   0   \n",
       "OR4F5                         0  ...                   0                   0   \n",
       "AL627309.1                    0  ...                   0                   0   \n",
       "AL627309.3                    0  ...                   0                   0   \n",
       "\n",
       "             TTTGTCACAAGAAGAG-1  TTTGTCACAGAAGCAC-1  TTTGTCAGTAGGGACT-1  \\\n",
       "MIR1302-2HG                   0                   0                   0   \n",
       "FAM138A                       0                   0                   0   \n",
       "OR4F5                         0                   0                   0   \n",
       "AL627309.1                    0                   0                   0   \n",
       "AL627309.3                    0                   0                   0   \n",
       "\n",
       "             TTTGTCAGTTCAGGCC-1  TTTGTCATCAACCAAC-1  TTTGTCATCACATGCA-1  \\\n",
       "MIR1302-2HG                   0                   0                   0   \n",
       "FAM138A                       0                   0                   0   \n",
       "OR4F5                         0                   0                   0   \n",
       "AL627309.1                    0                   0                   0   \n",
       "AL627309.3                    0                   0                   0   \n",
       "\n",
       "             TTTGTCATCCAAACAC-1  TTTGTCATCGCGTTTC-1  \n",
       "MIR1302-2HG                   0                   0  \n",
       "FAM138A                       0                   0  \n",
       "OR4F5                         0                   0  \n",
       "AL627309.1                    0                   0  \n",
       "AL627309.3                    0                   0  \n",
       "\n",
       "[5 rows x 6249 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# for each h5 file\n",
    "# df: data = count_matrix, columns = barcodes, index = features/name\n",
    "\n",
    "def parse_h5(file_path):\n",
    "    with h5py.File(file_path, 'r') as f:\n",
    "        data = f['matrix']['data'][()]\n",
    "        barcodes = f['matrix']['barcodes'][()].astype(str)\n",
    "        Genename = f['matrix']['features/name'][()].astype(str)\n",
    "        shape = f['matrix']['shape']\n",
    "        indices = f['matrix']['indices']\n",
    "        indptr = f['matrix']['indptr']\n",
    "        shape = f['matrix']['shape']\n",
    "        # Get the number of features and barcodes\n",
    "        num_features = shape[0]\n",
    "        num_barcodes = shape[1]\n",
    "\n",
    "        count_matrix = np.zeros((num_features, num_barcodes), dtype=int)\n",
    "\n",
    "        for i in range(num_barcodes):\n",
    "            barcode_data = data[indptr[i]:indptr[i+1]]\n",
    "            barcode_indices = indices[indptr[i]:indptr[i+1]]\n",
    "            count_matrix[barcode_indices, i] = barcode_data\n",
    "\n",
    "        df = pd.DataFrame(data=count_matrix, columns=barcodes, index=Genename)\n",
    "        return df\n",
    "\n",
    "df_test = parse_h5(os.path.join(os.getcwd(), 'GSM4339769_C141_filtered_feature_bc_matrix.h5'))\n",
    "df_test.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8ccd4ccf-550d-42fb-b8de-accb172dd4c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_mild samples: C144, C142, C141\n",
    "# df_severe samples: C146, C143, C145\n",
    "# df_control samples: C48, C49, C50, C51, C52, C53\n",
    "\n",
    "def df_T(df):\n",
    "    \"\"\"\n",
    "    Transpose a Pandas DataFrame while preserving both column names and index.\n",
    "\n",
    "    Parameters:\n",
    "    df (pd.DataFrame): Input DataFrame to be transposed.\n",
    "\n",
    "    Returns:\n",
    "    pd.DataFrame: Transposed DataFrame with column names and indexes preserved.\n",
    "    \"\"\"\n",
    "    transposed_df = df.transpose()\n",
    "    transposed_df.columns.name = 'Index'\n",
    "    \n",
    "    return transposed_df\n",
    "\n",
    "df_mild_T = []\n",
    "df_severe_T = []\n",
    "df_control_T = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a70b719b-ec59-48b4-ab8d-c72a1fb47650",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Index</th>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <th>FAM138A</th>\n",
       "      <th>OR4F5</th>\n",
       "      <th>AL627309.1</th>\n",
       "      <th>AL627309.3</th>\n",
       "      <th>AL627309.2</th>\n",
       "      <th>AL627309.4</th>\n",
       "      <th>AL732372.1</th>\n",
       "      <th>OR4F29</th>\n",
       "      <th>AC114498.1</th>\n",
       "      <th>...</th>\n",
       "      <th>BX072566.1</th>\n",
       "      <th>AL354822.1</th>\n",
       "      <th>AC023491.2</th>\n",
       "      <th>AC004556.1</th>\n",
       "      <th>AC233755.2</th>\n",
       "      <th>AC233755.1</th>\n",
       "      <th>AC240274.1</th>\n",
       "      <th>AC213203.1</th>\n",
       "      <th>FAM231C</th>\n",
       "      <th>nCoV</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGATGTCGG-1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGCTCATT-1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGTTGTCGT-1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCAATCCGAT-1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCAGGCGATA-1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 33539 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Index               MIR1302-2HG  FAM138A  OR4F5  AL627309.1  AL627309.3  \\\n",
       "AAACCTGAGATGTCGG-1            0        0      0           0           0   \n",
       "AAACCTGAGGCTCATT-1            0        0      0           0           0   \n",
       "AAACCTGAGTTGTCGT-1            0        0      0           0           0   \n",
       "AAACCTGCAATCCGAT-1            0        0      0           0           0   \n",
       "AAACCTGCAGGCGATA-1            0        0      0           0           0   \n",
       "\n",
       "Index               AL627309.2  AL627309.4  AL732372.1  OR4F29  AC114498.1  \\\n",
       "AAACCTGAGATGTCGG-1           0           0           0       0           0   \n",
       "AAACCTGAGGCTCATT-1           0           0           0       0           0   \n",
       "AAACCTGAGTTGTCGT-1           0           0           0       0           0   \n",
       "AAACCTGCAATCCGAT-1           0           0           0       0           0   \n",
       "AAACCTGCAGGCGATA-1           0           0           0       0           0   \n",
       "\n",
       "Index               ...  BX072566.1  AL354822.1  AC023491.2  AC004556.1  \\\n",
       "AAACCTGAGATGTCGG-1  ...           0           0           0           0   \n",
       "AAACCTGAGGCTCATT-1  ...           0           0           0           0   \n",
       "AAACCTGAGTTGTCGT-1  ...           0           0           0           0   \n",
       "AAACCTGCAATCCGAT-1  ...           0           0           0           0   \n",
       "AAACCTGCAGGCGATA-1  ...           0           0           0           0   \n",
       "\n",
       "Index               AC233755.2  AC233755.1  AC240274.1  AC213203.1  FAM231C  \\\n",
       "AAACCTGAGATGTCGG-1           0           0           0           0        0   \n",
       "AAACCTGAGGCTCATT-1           0           0           0           0        0   \n",
       "AAACCTGAGTTGTCGT-1           0           0           0           0        0   \n",
       "AAACCTGCAATCCGAT-1           0           0           0           0        0   \n",
       "AAACCTGCAGGCGATA-1           0           0           0           0        0   \n",
       "\n",
       "Index               nCoV  \n",
       "AAACCTGAGATGTCGG-1     0  \n",
       "AAACCTGAGGCTCATT-1     0  \n",
       "AAACCTGAGTTGTCGT-1     0  \n",
       "AAACCTGCAATCCGAT-1     0  \n",
       "AAACCTGCAGGCGATA-1     0  \n",
       "\n",
       "[5 rows x 33539 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_T_test = df_T(parse_h5(os.path.join(os.getcwd(), 'GSM4339769_C141_filtered_feature_bc_matrix.h5')))\n",
    "df_T_test.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "84d57b23-3ed7-4900-b9b4-aca6e451fe8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Iterate over the lists of filenames for each severity level and parse each file\n",
    "df_mild = []\n",
    "for file in mild:\n",
    "    df_mild.append(df_T(parse_h5(os.path.join(os.getcwd(), file))))\n",
    "\n",
    "df_severe = []\n",
    "for file in severe:\n",
    "    df_severe.append(df_T(parse_h5(os.path.join(os.getcwd(), file))))\n",
    "\n",
    "df_control = []\n",
    "for file in control:\n",
    "    df_control.append(df_T(parse_h5(os.path.join(os.getcwd(), file))))\n",
    "\n",
    "# Now you have lists of DataFrames for each severity level\n",
    "df_mild_1, df_mild_2, df_mild_3 = df_mild\n",
    "df_severe_1, df_severe_2, df_severe_3 = df_severe\n",
    "df_control_1, df_control_2, df_control_3, df_control_4, df_control_5, df_control_6 = df_control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "be745913-c53a-4801-81e2-c5b9982417f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the sample names for each severity level\n",
    "mild_samples = ['C144', 'C142', 'C141']\n",
    "severe_samples = ['C146', 'C143', 'C145']\n",
    "control_samples = ['C48', 'C49', 'C50', 'C51', 'C52', 'C53']\n",
    "\n",
    "# Function to add 'sample' column to a DataFrame with the sample values\n",
    "def add_sample_column(df, sample_names):\n",
    "    df['sample'] = sample_names\n",
    "    return df\n",
    "\n",
    "# Add 'sample' column to DataFrames in the lists\n",
    "df_mild = [add_sample_column(df, sample) for df, sample in zip(df_mild, mild_samples)]\n",
    "df_severe = [add_sample_column(df, sample) for df, sample in zip(df_severe, severe_samples)]\n",
    "df_control = [add_sample_column(df, sample) for df, sample in zip(df_control, control_samples)]\n",
    "\n",
    "# Now each DataFrame in df_mild, df_severe, and df_control has a 'sample' column added"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "513bffa3-286c-4272-b3e9-4b0b5efaa701",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Index</th>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <th>FAM138A</th>\n",
       "      <th>OR4F5</th>\n",
       "      <th>AL627309.1</th>\n",
       "      <th>AL627309.3</th>\n",
       "      <th>AL627309.2</th>\n",
       "      <th>AL627309.4</th>\n",
       "      <th>AL732372.1</th>\n",
       "      <th>OR4F29</th>\n",
       "      <th>AC114498.1</th>\n",
       "      <th>...</th>\n",
       "      <th>AL354822.1</th>\n",
       "      <th>AC023491.2</th>\n",
       "      <th>AC004556.1</th>\n",
       "      <th>AC233755.2</th>\n",
       "      <th>AC233755.1</th>\n",
       "      <th>AC240274.1</th>\n",
       "      <th>AC213203.1</th>\n",
       "      <th>FAM231C</th>\n",
       "      <th>nCoV</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGCTAGTGG-1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>C144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCAGCCTTTC-1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>C144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCAGCTGTGC-1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>C144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGGTCTAGTGT-1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>C144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGGTGACTCAT-1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>C144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAAGTCAAGGC-1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>C144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCACAGGCAGTA-1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>C144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTACTTCTT-1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>C144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTGACCAAG-1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>C144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCTCTGAGA-1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>C144</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3716 rows × 33540 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Index               MIR1302-2HG  FAM138A  OR4F5  AL627309.1  AL627309.3  \\\n",
       "AAACCTGAGCTAGTGG-1            0        0      0           0           0   \n",
       "AAACCTGCAGCCTTTC-1            0        0      0           0           0   \n",
       "AAACCTGCAGCTGTGC-1            0        0      0           0           0   \n",
       "AAACCTGGTCTAGTGT-1            0        0      0           0           0   \n",
       "AAACCTGGTGACTCAT-1            0        0      0           0           0   \n",
       "...                         ...      ...    ...         ...         ...   \n",
       "TTTGTCAAGTCAAGGC-1            0        0      0           0           0   \n",
       "TTTGTCACAGGCAGTA-1            0        0      0           0           0   \n",
       "TTTGTCAGTACTTCTT-1            0        0      0           0           0   \n",
       "TTTGTCAGTGACCAAG-1            0        0      0           0           0   \n",
       "TTTGTCATCTCTGAGA-1            0        0      0           0           0   \n",
       "\n",
       "Index               AL627309.2  AL627309.4  AL732372.1  OR4F29  AC114498.1  \\\n",
       "AAACCTGAGCTAGTGG-1           0           0           0       0           0   \n",
       "AAACCTGCAGCCTTTC-1           0           0           0       0           0   \n",
       "AAACCTGCAGCTGTGC-1           0           0           0       0           0   \n",
       "AAACCTGGTCTAGTGT-1           0           0           0       0           0   \n",
       "AAACCTGGTGACTCAT-1           0           0           0       0           0   \n",
       "...                        ...         ...         ...     ...         ...   \n",
       "TTTGTCAAGTCAAGGC-1           0           0           0       0           0   \n",
       "TTTGTCACAGGCAGTA-1           0           0           0       0           0   \n",
       "TTTGTCAGTACTTCTT-1           0           0           0       0           0   \n",
       "TTTGTCAGTGACCAAG-1           0           0           0       0           0   \n",
       "TTTGTCATCTCTGAGA-1           0           0           0       0           0   \n",
       "\n",
       "Index               ...  AL354822.1  AC023491.2  AC004556.1  AC233755.2  \\\n",
       "AAACCTGAGCTAGTGG-1  ...           0           0           0           0   \n",
       "AAACCTGCAGCCTTTC-1  ...           0           0           0           0   \n",
       "AAACCTGCAGCTGTGC-1  ...           0           0           0           0   \n",
       "AAACCTGGTCTAGTGT-1  ...           0           0           0           0   \n",
       "AAACCTGGTGACTCAT-1  ...           0           0           0           0   \n",
       "...                 ...         ...         ...         ...         ...   \n",
       "TTTGTCAAGTCAAGGC-1  ...           0           0           0           0   \n",
       "TTTGTCACAGGCAGTA-1  ...           0           0           1           0   \n",
       "TTTGTCAGTACTTCTT-1  ...           0           0           0           0   \n",
       "TTTGTCAGTGACCAAG-1  ...           0           0           0           0   \n",
       "TTTGTCATCTCTGAGA-1  ...           0           0           0           0   \n",
       "\n",
       "Index               AC233755.1  AC240274.1  AC213203.1  FAM231C  nCoV  sample  \n",
       "AAACCTGAGCTAGTGG-1           0           0           0        0     0    C144  \n",
       "AAACCTGCAGCCTTTC-1           0           0           0        0     0    C144  \n",
       "AAACCTGCAGCTGTGC-1           0           0           0        0     0    C144  \n",
       "AAACCTGGTCTAGTGT-1           0           0           0        0     0    C144  \n",
       "AAACCTGGTGACTCAT-1           0           0           0        0     0    C144  \n",
       "...                        ...         ...         ...      ...   ...     ...  \n",
       "TTTGTCAAGTCAAGGC-1           0           0           0        0     0    C144  \n",
       "TTTGTCACAGGCAGTA-1           0           0           0        0     0    C144  \n",
       "TTTGTCAGTACTTCTT-1           0           0           0        0     0    C144  \n",
       "TTTGTCAGTGACCAAG-1           0           0           0        0     0    C144  \n",
       "TTTGTCATCTCTGAGA-1           0           0           0        0     0    C144  \n",
       "\n",
       "[3716 rows x 33540 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_mild[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ac3342a6-8038-4006-a92c-f482f0f03876",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_columns_with_zeros(df_list, threshold=0.8):\n",
    "    filtered_dfs = []\n",
    "    \n",
    "    for df in df_list:\n",
    "        # Calculate the threshold for deletion\n",
    "        threshold_count = int(threshold * len(df))\n",
    "        \n",
    "        # Filter out columns with more than the threshold count of zeros\n",
    "        cols_to_keep = df.columns[(df == 0).sum() < threshold_count]\n",
    "        \n",
    "        # Create a new DataFrame with columns having less than the threshold count of zeros\n",
    "        new_df = df[cols_to_keep]\n",
    "        filtered_dfs.append(new_df)\n",
    "    \n",
    "    return filtered_dfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "62036db1-ce5a-4e6b-b608-be90cfaa809b",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_mild = filter_columns_with_zeros(df_mild, threshold=0.5)\n",
    "filtered_severe = filter_columns_with_zeros(df_mild, threshold=0.5)\n",
    "filtered_control = filter_columns_with_zeros(df_control, threshold=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4344726a-ab52-4077-b1c4-668fe02d85b5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3716, 19)\n",
      "(10269, 42)\n",
      "(6249, 480)\n",
      "\n",
      "\n",
      "(3716, 19)\n",
      "(10269, 42)\n",
      "(6249, 480)\n",
      "\n",
      "\n",
      "(2879, 368)\n",
      "(10366, 342)\n",
      "(7732, 169)\n",
      "(3920, 159)\n",
      "(11115, 500)\n",
      "(8972, 526)\n"
     ]
    }
   ],
   "source": [
    "print(filtered_mild[0].shape)\n",
    "print(filtered_mild[1].shape)\n",
    "print(filtered_mild[2].shape)\n",
    "print('\\n')\n",
    "print(filtered_severe[0].shape)\n",
    "print(filtered_severe[1].shape)\n",
    "print(filtered_severe[2].shape)\n",
    "print('\\n')\n",
    "print(filtered_control[0].shape)\n",
    "print(filtered_control[1].shape)\n",
    "print(filtered_control[2].shape)\n",
    "print(filtered_control[3].shape)\n",
    "print(filtered_control[4].shape)\n",
    "print(filtered_control[5].shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69745eb2-11f6-4b65-8752-4ddc5c80decd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Combine all filtered DataFrames for each category\n",
    "all_filtered = filtered_mild + filtered_severe + filtered_control\n",
    "\n",
    "# Get unique column names across all DataFrames\n",
    "unique_columns = set().union(*[df.columns.difference(['nCoV', 'sample']) for df in all_filtered])\n",
    "\n",
    "# Ensure each DataFrame has the same set of columns\n",
    "for df_list in [filtered_mild, filtered_severe, filtered_control]:\n",
    "    for df in df_list:\n",
    "        missing_cols = unique_columns.difference(df.columns)\n",
    "        for col in missing_cols:\n",
    "            df[col] = 0\n",
    "\n",
    "# Verify the shapes after adding missing columns\n",
    "for df_list in [filtered_mild, filtered_severe, filtered_control]:\n",
    "    for df in df_list:\n",
    "        print(df.shape)\n",
    "    print('\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "662b682b-0239-46b3-9d3f-59cc31142b8c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3716, 750)\n",
      "(10269, 750)\n",
      "(6249, 750)\n",
      "\n",
      "\n",
      "(3716, 750)\n",
      "(10269, 750)\n",
      "(6249, 750)\n",
      "\n",
      "\n",
      "(2879, 750)\n",
      "(10366, 750)\n",
      "(7732, 750)\n",
      "(3920, 750)\n",
      "(11115, 750)\n",
      "(8972, 750)\n"
     ]
    }
   ],
   "source": [
    "print(filtered_mild[0].shape)\n",
    "print(filtered_mild[1].shape)\n",
    "print(filtered_mild[2].shape)\n",
    "print('\\n')\n",
    "print(filtered_severe[0].shape)\n",
    "print(filtered_severe[1].shape)\n",
    "print(filtered_severe[2].shape)\n",
    "print('\\n')\n",
    "print(filtered_control[0].shape)\n",
    "print(filtered_control[1].shape)\n",
    "print(filtered_control[2].shape)\n",
    "print(filtered_control[3].shape)\n",
    "print(filtered_control[4].shape)\n",
    "print(filtered_control[5].shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "5930276d-9e6d-4cf2-9f64-5c8526f788d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the directory path to write the CSV files\n",
    "directory = os.getcwd()\n",
    "\n",
    "# Write each DataFrame to a separate CSV file\n",
    "for idx, df_list in enumerate([filtered_mild, filtered_severe, filtered_control]):\n",
    "    for i, df in enumerate(df_list):\n",
    "        file_path = os.path.join(directory, f\"filtered_{idx}_df_{i}.csv\")\n",
    "        df.to_csv(file_path, index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee9957c5-724f-4363-a88f-159c5c8fc181",
   "metadata": {},
   "source": [
    "## Part II: Annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6c9a9264-e241-4564-a160-878ad253f8ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import csv\n",
    "import pandas as pd\n",
    "import matplotlib\n",
    "import sklearn\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from sklearn import preprocessing\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.decomposition import PCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ee3d28f7-ec77-42aa-b4ab-aecaf9396bc3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['file_description.rtf',\n",
       " 'GSM4475052_C149_filtered_feature_bc_matrix.h5',\n",
       " 'GSM4339774_C146_filtered_feature_bc_matrix.h5',\n",
       " 'GSM4475049_C52_filtered_feature_bc_matrix.h5',\n",
       " 'GSM4339772_C144_filtered_feature_bc_matrix.h5',\n",
       " 'GSM4475053_C152_filtered_feature_bc_matrix.h5',\n",
       " 'GSM4475051_C148_filtered_feature_bc_matrix.h5',\n",
       " 'GSM4339770_C142_filtered_feature_bc_matrix.h5',\n",
       " 'GSM4339769_C141_filtered_feature_bc_matrix.h5',\n",
       " 'GSM4339771_C143_filtered_feature_bc_matrix.h5',\n",
       " 'GSM4475048_C51_filtered_feature_bc_matrix.h5',\n",
       " 'GSM4339773_C145_filtered_feature_bc_matrix.h5',\n",
       " 'GSM4475050_C100_filtered_feature_bc_matrix.h5',\n",
       " '.ipynb_checkpoints',\n",
       " 'file_description.txt',\n",
       " 'ML_Script_1_data_preprocessing.ipynb',\n",
       " 'all.cell.annotation.meta.txt',\n",
       " 'filtered_0_df_0.csv',\n",
       " 'filtered_0_df_1.csv',\n",
       " 'filtered_0_df_2.csv',\n",
       " 'filtered_1_df_0.csv',\n",
       " 'filtered_1_df_1.csv',\n",
       " 'filtered_1_df_2.csv',\n",
       " 'filtered_2_df_0.csv',\n",
       " 'filtered_2_df_1.csv',\n",
       " 'filtered_2_df_2.csv',\n",
       " 'filtered_2_df_3.csv',\n",
       " 'filtered_2_df_4.csv',\n",
       " 'filtered_2_df_5.csv',\n",
       " 'annot_0_df_0.csv',\n",
       " 'annot_0_df_1.csv',\n",
       " 'annot_0_df_2.csv',\n",
       " 'annot_1_df_0.csv',\n",
       " 'annot_1_df_1.csv',\n",
       " 'annot_1_df_2.csv',\n",
       " 'annot_2_df_0.csv',\n",
       " 'annot_2_df_1.csv',\n",
       " 'annot_2_df_2.csv',\n",
       " 'annot_2_df_3.csv',\n",
       " 'annot_2_df_4.csv',\n",
       " 'annot_2_df_5.csv']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.listdir()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b4e93a03-c9c9-42dd-9677-3e1cb4c76558",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3716, 750)\n",
      "(10269, 750)\n",
      "(6249, 750)\n",
      "(3716, 750)\n",
      "(10269, 750)\n",
      "(6249, 750)\n",
      "(2879, 750)\n",
      "(10366, 750)\n",
      "(7732, 750)\n",
      "(3920, 750)\n",
      "(11115, 750)\n",
      "(8972, 750)\n"
     ]
    }
   ],
   "source": [
    "# Define the directory where the CSV files are stored\n",
    "directory = os.getcwd()\n",
    "\n",
    "# Initialize lists to store the DataFrames\n",
    "filtered_mild = []\n",
    "filtered_severe = []\n",
    "filtered_control = []\n",
    "\n",
    "# Read the CSV files back into the original variables\n",
    "for filename in os.listdir(directory):\n",
    "    if filename.endswith(\".csv\"):\n",
    "        if \"filtered_0_\" in filename:\n",
    "            df = pd.read_csv(os.path.join(directory, filename))\n",
    "            filtered_mild.append(df.set_index(df.columns[0]))  # Set the index\n",
    "        elif \"filtered_1_\" in filename:\n",
    "            df = pd.read_csv(os.path.join(directory, filename))\n",
    "            filtered_severe.append(df.set_index(df.columns[0]))  # Set the index\n",
    "        elif \"filtered_2_\" in filename:\n",
    "            df = pd.read_csv(os.path.join(directory, filename))\n",
    "            filtered_control.append(df.set_index(df.columns[0]))  # Set the index\n",
    "\n",
    "\n",
    "# Verify the DataFrames have been read back correctly\n",
    "for df in filtered_mild + filtered_severe + filtered_control:\n",
    "    print(df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f97ec509-e86d-445b-a1b2-7b14a1201973",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3716, 750)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_mild[0].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9fec1446-607f-4ae8-95f2-bfaebe10df42",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sample</th>\n",
       "      <th>sample_new</th>\n",
       "      <th>group</th>\n",
       "      <th>disease</th>\n",
       "      <th>hasnCoV</th>\n",
       "      <th>cluster</th>\n",
       "      <th>celltype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCTGAGACACTAA_1</td>\n",
       "      <td>C51</td>\n",
       "      <td>HC1</td>\n",
       "      <td>HC</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>3</td>\n",
       "      <td>Macrophages</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCTGAGGAGTACC_1</td>\n",
       "      <td>C51</td>\n",
       "      <td>HC1</td>\n",
       "      <td>HC</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>3</td>\n",
       "      <td>Macrophages</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCTGAGGATATAC_1</td>\n",
       "      <td>C51</td>\n",
       "      <td>HC1</td>\n",
       "      <td>HC</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>3</td>\n",
       "      <td>Macrophages</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCTGAGGTCATCT_1</td>\n",
       "      <td>C51</td>\n",
       "      <td>HC1</td>\n",
       "      <td>HC</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>3</td>\n",
       "      <td>Macrophages</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCTGCACGGATAG_1</td>\n",
       "      <td>C51</td>\n",
       "      <td>HC1</td>\n",
       "      <td>HC</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>5</td>\n",
       "      <td>Macrophages</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   ID sample sample_new group disease hasnCoV  cluster  \\\n",
       "0  AAACCTGAGACACTAA_1    C51        HC1    HC       N       N        3   \n",
       "1  AAACCTGAGGAGTACC_1    C51        HC1    HC       N       N        3   \n",
       "2  AAACCTGAGGATATAC_1    C51        HC1    HC       N       N        3   \n",
       "3  AAACCTGAGGTCATCT_1    C51        HC1    HC       N       N        3   \n",
       "4  AAACCTGCACGGATAG_1    C51        HC1    HC       N       N        5   \n",
       "\n",
       "      celltype  \n",
       "0  Macrophages  \n",
       "1  Macrophages  \n",
       "2  Macrophages  \n",
       "3  Macrophages  \n",
       "4  Macrophages  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add cell annotation & file annotation\n",
    "all_cell_annot = pd.read_csv(os.path.join(os.getcwd(), 'all.cell.annotation.meta.txt'), sep='\\t')\n",
    "all_cell_annot.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "99d28f17-b653-4282-bc52-407ce237bc86",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MTRNR2L12</th>\n",
       "      <th>HBB</th>\n",
       "      <th>MTRNR2L8</th>\n",
       "      <th>TAOK1</th>\n",
       "      <th>FTL</th>\n",
       "      <th>MT-ND1</th>\n",
       "      <th>MT-ND2</th>\n",
       "      <th>MT-CO1</th>\n",
       "      <th>MT-CO2</th>\n",
       "      <th>MT-ATP8</th>\n",
       "      <th>...</th>\n",
       "      <th>HERC5</th>\n",
       "      <th>CES1</th>\n",
       "      <th>EIF2S2</th>\n",
       "      <th>HLA-DRB1</th>\n",
       "      <th>RPL38</th>\n",
       "      <th>CXCL10</th>\n",
       "      <th>ACTG1</th>\n",
       "      <th>CANX</th>\n",
       "      <th>HNRNPA2B1</th>\n",
       "      <th>CSTB</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGCTAGTGG-1</th>\n",
       "      <td>0</td>\n",
       "      <td>654</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCAGCCTTTC-1</th>\n",
       "      <td>105</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>36</td>\n",
       "      <td>0</td>\n",
       "      <td>269</td>\n",
       "      <td>113</td>\n",
       "      <td>1066</td>\n",
       "      <td>1271</td>\n",
       "      <td>523</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCAGCTGTGC-1</th>\n",
       "      <td>35</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>58</td>\n",
       "      <td>254</td>\n",
       "      <td>283</td>\n",
       "      <td>244</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGGTCTAGTGT-1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGGTGACTCAT-1</th>\n",
       "      <td>192</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>29</td>\n",
       "      <td>1</td>\n",
       "      <td>33</td>\n",
       "      <td>44</td>\n",
       "      <td>132</td>\n",
       "      <td>174</td>\n",
       "      <td>47</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 750 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    MTRNR2L12  HBB  MTRNR2L8  TAOK1  FTL  MT-ND1  MT-ND2  \\\n",
       "Unnamed: 0                                                                 \n",
       "AAACCTGAGCTAGTGG-1          0  654         0      5    5       0       1   \n",
       "AAACCTGCAGCCTTTC-1        105    1         9     36    0     269     113   \n",
       "AAACCTGCAGCTGTGC-1         35    0         4     18    0     100      58   \n",
       "AAACCTGGTCTAGTGT-1          0    0         0      7   13       1       1   \n",
       "AAACCTGGTGACTCAT-1        192    0        14     29    1      33      44   \n",
       "\n",
       "                    MT-CO1  MT-CO2  MT-ATP8  ...  HERC5  CES1  EIF2S2  \\\n",
       "Unnamed: 0                                   ...                        \n",
       "AAACCTGAGCTAGTGG-1       6      10        0  ...      0     0       0   \n",
       "AAACCTGCAGCCTTTC-1    1066    1271      523  ...      0     0       0   \n",
       "AAACCTGCAGCTGTGC-1     254     283      244  ...      0     0       0   \n",
       "AAACCTGGTCTAGTGT-1       7      10        2  ...      0     0       0   \n",
       "AAACCTGGTGACTCAT-1     132     174       47  ...      0     0       0   \n",
       "\n",
       "                    HLA-DRB1  RPL38  CXCL10  ACTG1  CANX HNRNPA2B1  CSTB  \n",
       "Unnamed: 0                                                                \n",
       "AAACCTGAGCTAGTGG-1         0      0       0      0     0         0     0  \n",
       "AAACCTGCAGCCTTTC-1         0      0       0      0     0         0     0  \n",
       "AAACCTGCAGCTGTGC-1         0      0       0      0     0         0     0  \n",
       "AAACCTGGTCTAGTGT-1         0      0       0      0     0         0     0  \n",
       "AAACCTGGTGACTCAT-1         0      0       0      0     0         0     0  \n",
       "\n",
       "[5 rows x 750 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_mild[0].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ed08610d-c37d-4efe-a0f1-57a1de78a9b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sample</th>\n",
       "      <th>sample_new</th>\n",
       "      <th>group</th>\n",
       "      <th>disease</th>\n",
       "      <th>hasnCoV</th>\n",
       "      <th>cluster</th>\n",
       "      <th>celltype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCTGAGACACTAA_1</td>\n",
       "      <td>C51</td>\n",
       "      <td>HC1</td>\n",
       "      <td>HC</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>3</td>\n",
       "      <td>Macrophages</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCTGAGGAGTACC_1</td>\n",
       "      <td>C51</td>\n",
       "      <td>HC1</td>\n",
       "      <td>HC</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>3</td>\n",
       "      <td>Macrophages</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCTGAGGATATAC_1</td>\n",
       "      <td>C51</td>\n",
       "      <td>HC1</td>\n",
       "      <td>HC</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>3</td>\n",
       "      <td>Macrophages</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCTGAGGTCATCT_1</td>\n",
       "      <td>C51</td>\n",
       "      <td>HC1</td>\n",
       "      <td>HC</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>3</td>\n",
       "      <td>Macrophages</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCTGCACGGATAG_1</td>\n",
       "      <td>C51</td>\n",
       "      <td>HC1</td>\n",
       "      <td>HC</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>5</td>\n",
       "      <td>Macrophages</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   ID sample sample_new group disease hasnCoV  cluster  \\\n",
       "0  AAACCTGAGACACTAA_1    C51        HC1    HC       N       N        3   \n",
       "1  AAACCTGAGGAGTACC_1    C51        HC1    HC       N       N        3   \n",
       "2  AAACCTGAGGATATAC_1    C51        HC1    HC       N       N        3   \n",
       "3  AAACCTGAGGTCATCT_1    C51        HC1    HC       N       N        3   \n",
       "4  AAACCTGCACGGATAG_1    C51        HC1    HC       N       N        5   \n",
       "\n",
       "      celltype  \n",
       "0  Macrophages  \n",
       "1  Macrophages  \n",
       "2  Macrophages  \n",
       "3  Macrophages  \n",
       "4  Macrophages  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_cell_annot.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "3b0e7a47-c809-40b5-8da0-e5cf98bec593",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['AAACCTGAGACACTAA-1' 'AAACCTGAGGAGTACC-1' 'AAACCTGAGGATATAC-1' ...\n",
      " 'TTTGTCACAACACCCG-13' 'TTTGTCAGTCTGCAAT-13' 'TTTGTCAGTGTTGGGA-13']\n"
     ]
    }
   ],
   "source": [
    "print(all_cell_annot['ID'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "a5dc0730-3457-4470-9af2-febb39f91eab",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['AAACCTGAGCTAGTGG-1', 'AAACCTGCAGCCTTTC-1', 'AAACCTGCAGCTGTGC-1',\n",
      "       'AAACCTGGTCTAGTGT-1', 'AAACCTGGTGACTCAT-1', 'AAACCTGTCGTCCGTT-1',\n",
      "       'AAACGGGAGACGCTTT-1', 'AAACGGGAGATCGGGT-1', 'AAACGGGAGCGTCTAT-1',\n",
      "       'AAACGGGCAATCGAAA-1',\n",
      "       ...\n",
      "       'TTTGGTTTCACATAGC-1', 'TTTGGTTTCGCTTGTC-1', 'TTTGGTTTCGTAGATC-1',\n",
      "       'TTTGGTTTCGTAGGAG-1', 'TTTGGTTTCTCCAACC-1', 'TTTGTCAAGTCAAGGC-1',\n",
      "       'TTTGTCACAGGCAGTA-1', 'TTTGTCAGTACTTCTT-1', 'TTTGTCAGTGACCAAG-1',\n",
      "       'TTTGTCATCTCTGAGA-1'],\n",
      "      dtype='object', name='Unnamed: 0', length=3716)\n"
     ]
    }
   ],
   "source": [
    "print(filtered_mild[0].index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a67685b-7a83-4304-9395-bd8e693dac10",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def extract_celltypes(df, annot):\n",
    "#     celltypes = []\n",
    "#     for barcode in df.index.astype(str):\n",
    "#         barcode_prefix = barcode.split('-')[0]\n",
    "#         if barcode_prefix in annot['ID'].astype(str).str.split('-').str[0].values:\n",
    "#             celltype = annot.loc[annot['ID'].str.split('-').str[0] == barcode_prefix, 'celltype'].values[0]\n",
    "#             celltypes.append(celltype)\n",
    "#         else:\n",
    "#             celltypes.append(None)\n",
    "#     return celltypes\n",
    "\n",
    "# # Append extracted cell types as a new column 'celltypes' to the filtered DataFrames\n",
    "# for df in filtered_mild:\n",
    "#     df['celltype'] = extract_celltypes(df, all_cell_annot)\n",
    "\n",
    "# for df in filtered_severe:\n",
    "#     df['celltype'] = extract_celltypes(df, all_cell_annot)\n",
    "\n",
    "# for df in filtered_control:\n",
    "#     df['celltype'] = extract_celltypes(df, all_cell_annot)\n",
    "\n",
    "# filtered_mild[0].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "31a64ec6-3a30-429b-9165-e9cb559a740a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "648\n"
     ]
    }
   ],
   "source": [
    "# Create lists of keys for annotation and DataFrame\n",
    "key_annot = [val.split('_')[0] for val in all_cell_annot['ID'].values]\n",
    "key_df = [val.split('-')[0] for val in filtered_mild[0].index.values]\n",
    "\n",
    "# Calculate the length of the intersection of sets\n",
    "print(len(set(key_annot) & set(key_df)))\n",
    "\n",
    "# not all cells can be annotated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "48b15618-8ced-45b4-ad2d-3e3eb085e00e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def filter_matched_rows(df, annot):\n",
    "#     to_keep = []\n",
    "#     # Create lists of keys for annotation and DataFrame\n",
    "#     key_annot = [val.split('_')[0] for val in annot['ID'].values]\n",
    "#     key_df = [val.split('-')[0] for val in df.index.values]\n",
    "\n",
    "#     for key in (set(key_annot) & set(key_df)):\n",
    "#         to_keep.extend(df[df.index.str.startswith(key)].index)\n",
    "\n",
    "#     filtered_df = df.loc[to_keep]\n",
    "#     return filtered_df\n",
    "\n",
    "# def map_celltypes(df, annot):\n",
    "#     celltypes = []\n",
    "#     for barcode in df.index.astype(str):\n",
    "#         barcode_prefix = barcode.split('-')[0]\n",
    "#         if barcode_prefix in annot['ID'].astype(str).str.split('_').str[0].values:\n",
    "#             celltype = annot.loc[annot['ID'].str.split('_').str[0] == barcode_prefix, 'celltype'].values[0]\n",
    "#             celltypes.append(celltype)\n",
    "#         else:\n",
    "#             celltypes.append(None)\n",
    "    \n",
    "#     return celltypes\n",
    "\n",
    "# filtered_lists = [filtered_mild, filtered_severe, filtered_control]\n",
    "\n",
    "# for filtered_list in filtered_lists:\n",
    "#     for df in filtered_list:\n",
    "#         # Step 1: Filter the DataFrame to keep rows with matching rows in all_cell_annot\n",
    "#         filtered_df = filter_matched_rows(df, all_cell_annot)\n",
    "        \n",
    "#         # Step 2: Map the cell types for the filtered DataFrame\n",
    "#         celltypes = map_celltypes(filtered_df, all_cell_annot)\n",
    "        \n",
    "#         # Assign the mapped cell types as a new column 'celltype' in the filtered DataFrame\n",
    "#         filtered_df['celltype'] = celltypes\n",
    "        \n",
    "#         # Display the updated DataFrame\n",
    "#         print(filtered_df.shape)\n",
    "\n",
    "# # Define the directory path to write the CSV files\n",
    "# directory = os.getcwd()\n",
    "\n",
    "# # Write each DataFrame to a separate CSV file\n",
    "# for idx, df_list in enumerate(filtered_lists):\n",
    "#     for i, df in enumerate(df_list):\n",
    "#         file_path = os.path.join(directory, f\"annot_{idx}_df_{i}.csv\")\n",
    "#         df.to_csv(file_path, index=True)\n",
    "\n",
    "## final csv shape not as expected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be676940-9c34-440e-9ef9-8fa163c590c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape of filtered DataFrame 0.0: (648, 751)\n",
      "Shape of filtered DataFrame 0.1: (3972, 751)\n",
      "Shape of filtered DataFrame 0.2: (3758, 751)\n",
      "Shape of filtered DataFrame 1.0: (648, 751)\n",
      "Shape of filtered DataFrame 1.1: (3972, 751)\n",
      "Shape of filtered DataFrame 1.2: (3758, 751)\n",
      "Shape of filtered DataFrame 2.0: (2124, 751)\n",
      "Shape of filtered DataFrame 2.1: (8310, 751)\n",
      "Shape of filtered DataFrame 2.2: (3279, 751)\n",
      "Shape of filtered DataFrame 2.3: (1907, 751)\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def filter_matched_rows(df, annot):\n",
    "    to_keep = []\n",
    "    # Create lists of keys for annotation and DataFrame\n",
    "    key_annot = [val.split('_')[0] for val in annot['ID'].values]\n",
    "    key_df = [val.split('-')[0] for val in df.index.values]\n",
    "\n",
    "    for key in (set(key_annot) & set(key_df)):\n",
    "        to_keep.extend(df[df.index.str.startswith(key)].index)\n",
    "\n",
    "    filtered_df = df.loc[to_keep]\n",
    "    return filtered_df\n",
    "\n",
    "def map_celltypes(df, annot):\n",
    "    celltypes = []\n",
    "    for barcode in df.index.astype(str):\n",
    "        barcode_prefix = barcode.split('-')[0]\n",
    "        if barcode_prefix in annot['ID'].astype(str).str.split('_').str[0].values:\n",
    "            celltype = annot.loc[annot['ID'].str.split('_').str[0] == barcode_prefix, 'celltype'].values[0]\n",
    "            celltypes.append(celltype)\n",
    "        else:\n",
    "            celltypes.append(None)\n",
    "    \n",
    "    return celltypes\n",
    "\n",
    "filtered_lists = [filtered_mild, filtered_severe, filtered_control]\n",
    "\n",
    "for i, filtered_list in enumerate(filtered_lists):\n",
    "    for j, df in enumerate(filtered_list):\n",
    "        # Step 1: Filter the DataFrame to keep rows with matching rows in all_cell_annot\n",
    "        filtered_df = filter_matched_rows(df, all_cell_annot)\n",
    "        \n",
    "        # Step 2: Map the cell types for the filtered DataFrame\n",
    "        celltypes = map_celltypes(filtered_df, all_cell_annot)\n",
    "        \n",
    "        # Assign the mapped cell types as a new column 'celltype' in the filtered DataFrame\n",
    "        filtered_df['celltype'] = celltypes\n",
    "        \n",
    "        # Display the updated DataFrame shape\n",
    "        print(f\"Shape of filtered DataFrame {i}.{j}: {filtered_df.shape}\")\n",
    "        \n",
    "        # Write the filtered DataFrame to a CSV file\n",
    "        filtered_df.to_csv(f'annot_{i}_df_{j}.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28daa1ce-b4fd-48ed-8286-b2227ad8dfe0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
